### PR TITLE
Middleware move permanently fix

### LINKED ISSUE #808: [🐛 Bug]: 301 Moved Permanently from middleware ended up with 307 Temporary Redirect
### next-on-pages environment related information

System:
        Platform: linux
        Arch: x64
        Version: #1 SMP Fri Mar 29 23:14:13 UTC 2024
        CPU: (16) x64 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
        Memory: 15 GB
        Shell: /bin/bash
Package Manager Used: npm (9.6.3)

Relevant Packages:
        @cloudflare/next-on-pages: 1.11.3
        vercel: 34.2.4
        next: 14.1.0

### Description

Cloudflare next-on-pages do 307 redirect while it has to be 301. 

### Reproduction

I have created a minimal reproducible example in this [repo](https://github.com/strofimo/cf-middleware-move-permanently)

## How to set it up

```sh
git clone https://github.com/strofimo/cf-middleware-move-permanently.git
cd cf-middleware-move-permanently
npm i
```

### Run NextJS Dev to see what are the expected redirect result
```sh
npm run dev
# accessing http://localhost:3000 lead to 301 Moved Permanently
# ▲ Next.js 14.1.0
# - Local:        http://localhost:3000
```

This is the expected behave for a next js app.

### Run Cloudflare pages preview using wrangler
```sh
npm run preview
# accessing http://localhost:8788 lead to 307 Temporary Redirect
# [wrangler:inf] Ready on http://localhost:8788
```

### Pages Deployment Method

None

### Pages Deployment ID

_No response_

### Additional Information

_No response_

### Would you like to help?

- [x] Would you like to help fixing this bug?

### PR DESCRIPTION
**Pull request by strofimo**
Suggested fix for issue [#808 ](https://github.com/cloudflare/next-on-pages/issues/808) where middleware redirect status not saving which leads to 307 redirect.

fixes #808

### CODE REVIEW

**Review by dario-piotrowicz — APPROVED** [review]
@strofimo this looks great to me! 🚀 

thanks so much for the issue and the PR! and the minimal repro that made validating this very straightforward! I really appreciate it ❤️ 

Ideally I would love to have an e2e test for this as it does look like something that can unintentionally regress in the future

If you'd be willing to create an e2e test for this that's be amazing! 🤩 
if not I can look into doing it later 🙂 (it definitely doesn't need to block this PR)

PS: writing e2es here is a bit painful... so don't feel like you have to, it's totally ok and understandable if you don't want to 😅👍 , just let me know 🙂 (as I'll wait for the e2e before merging the PR if you'd like to create one)

**Review by dario-piotrowicz — APPROVED** [review]

**Comment by dario-piotrowicz** [comment]
> Unfortunatelly the original fix was too greedy which cause the issues with rewrites.

I see, I'd love to see those rewrites since the ones we have in our e2es were not failing, so it'd be good to have more coverage there

The updated version looks even better thanks 😄👍

**Comment by dario-piotrowicz** [comment]
> Thanks a lot for quick turnaround!

thank you for the awesome contribution! it's really appreciated! 🫶

**Comment by dario-piotrowicz** [comment]
@strofimo could you run `npm run prettier:fix` to get the code formatted? 🙂 🙏

**Comment by strofimo** [comment]
> @strofimo could you run `npm run prettier:fix` to get the code formatted? 🙂 🙏

Ah sorry, missed it. Done!

**Comment by strofimo** [comment]
> > Unfortunatelly the original fix was too greedy which cause the issues with rewrites.
> 
> I see, I'd love to see those rewrites since the ones we have in our e2es were not failing, so it'd be good to have more coverage there
> 
> The updated version looks even better thanks 😄👍

Yeah we found it when tested the fix on our codebase. I will try to come up with simple test case if its possible.

**Pull request merged**
